File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:

        res = []

        if not digits:
            return res
        
        numMap = {"2":"abc", "3":"def", "4":"ghi", "5":"jkl", "6":"mno", "7":"pqrs", "8":"tuv", "9":"wxyz"}

        def dfs(current, i):

            if i == len(digits):
                res.append(current)
                return

            letters = numMap[digits[i]]

            for letter in letters:
                dfs(current+letter, i+1)

        dfs("", 0)
        
        return res

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {

    static final String[] letterArr = {"", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    public List<String> letterCombinations(String digits) {

        List<String> res = new ArrayList<String>();         

        if (digits.length() == 0) return res;

        dfs(res, new StringBuilder(), 0, digits);

        return res;
    }

    public void dfs(List<String> res, StringBuilder current, int index, String digits){

        if(index == digits.length()){
            res.add(current.toString());
            return;
        }

        String currStr = letterArr[digits.charAt(index)-'1'];

        for(int i=0; i<currStr.length(); i++){
            current.append(currStr.charAt(i));
            dfs(res, current, index+1, digits);
            current.deleteCharAt(current.length()-1);
        }
    }
}